### DISCUSSION TITLE
macOS 12.3 and python

### DISCUSSION THREAD

**Discussion by skovatch** [Ascoware/get-iplayer-automator#374] [Announcements]
In the first beta of macOS 12.3, Apple removes python 2.x from your hard drive. Because ITV downloads and youtube-dl currently rely on '/usr/bin/python' being there, Get iPlayer Automator will stop working when you update to 12.3.

So, to deal with this problem, I decided it was time to bundle a python distribution into GiA. This removes all dependency on the OS, except for some standard Unix libraries that are always available.

I have tested this on an M1 and Intel laptop running macOS Monterey. I would appreciate it if someone would try it out on Big Sur (12.1) or earlier, since I don't have access to hardware with the older OS on it.

[Download it here](https://drive.google.com/drive/folders/15-PmyvvmmPDxe8sAHIM-94U-9VJ0S-UW?usp=sharing)

**Comment by jeanpaulz85** [comment]
I have Big Sur, it did not work here is the log

Get iPlayer Automator 1.21.14 Initialized.
GetITVShows: ITV Cache Update Starting 
<AppController: 0x7fe99cf52b80>: Updating Programme Index Feeds...

/Applications/Get iPlayer Automator.app/Contents/Resources/get_iplayer/perl/bin/get_iplayer
-e14400
--type=tv
--refresh
--nopurge
--profile-dir=/Users/jean-paulzerbib/Library/Application Support/Get iPlayer Automator
.*
INFO: Indexing tv programmes (concurrent)
GetITVShows (Info): Episodes: 5217 Today Programmes: 480 
GetITVShows: Update Finished
INFO: Added 0 tv programmes to cache
INFO: 888 matching programmes
<AppController: 0x7fe99cf52b80>: Index Updated.


INFO: Loading proxy settings...
INFO: No proxy to load
INFO: Proxy load complete.
INFO: No proxy will be used
AppController: Starting Downloads


Downloading Show 1/1:

Downloading Trigger Point: Series 1
INFO: Preparing Request for Auth Info
INFO: Requesting Metadata.
DEBUG: Metadata response status code: 200
INFO: Metadata processed.
DEBUG: youtube-dl args:["/Applications/Get iPlayer Automator.app/Contents/Resources/youtube-dl", "https://www.itv.com/hub/trigger-point/10a0591a0002", "--user-agent", "Mozilla/5.0", "-f", "mp4/best", "-o", "/Users/jean-paulzerbib/Movies/TV Shows/Trigger Point/Trigger Point.s01e02.Episode 2.%(ext)s", "--write-sub", "-k", "--verbose"]
DEBUG: youtube-dl environment: ["SSL_CERT_FILE": "/Applications/Get iPlayer Automator.app/Contents/Resources/python/lib/python3.11/site-packages/certifi/cacert.pem", "PATH": "/Applications/Get iPlayer Automator.app/Contents/MacOS:/Applications/Get iPlayer Automator.app/Contents/Resources/python:/Applications/Get iPlayer Automator.app/Contents/Resources/get_iplayer/utils/bin", "PYTHONPATH": "/Applications/Get iPlayer Automator.app/Contents/Resources/python"]
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Applications/Get iPlayer Automator.app/Contents/Resources/youtube-dl/__main__.py", line 16, in <module>
  File "/Applications/Get iPlayer Automator.app/Contents/Resources/youtube-dl/youtube_dl/__init__.py", line 15, in <module>
  File "/Applications/Get iPlayer Automator.app/Contents/Resources/youtube-dl/youtube_dl/options.py", line 8, in <module>
  File "/Applications/Get iPlayer Automator.app/Contents/Resources/youtube-dl/youtube_dl/downloader/__init__.py", line 3, in <module>
  File "/Applications/Get iPlayer Automator.app/Contents/Resources/youtube-dl/youtube_dl/downloader/common.py", line 10, in <module>
  File "/Applications/Get iPlayer Automator.app/Contents/Resources/youtube-dl/youtube_dl/utils.py", line 30, in <module>
  File "/Applications/Get iPlayer Automator.app/Contents/Resources/python/lib/python3.11/ssl.py", line 99, in <module>
    import _ssl             # if we can't import it, let the error propagate
    ^^^^^^^^^^^
ImportError: dlopen(/Applications/Get iPlayer Automator.app/Contents/Resources/python/lib/python3.11/lib-dynload/_ssl.cpython-311-darwin.so, 2): Library not loaded: @loader_path/libssl.3.dylib
  Referenced from: /Applications/Get iPlayer Automator.app/Contents/Resources/python/lib/python3.11/lib-dynload/_ssl.cpython-311-darwin.so
  Reason: no suitable image found.  Did find:
	/Applications/Get iPlayer Automator.app/Contents/Resources/python/lib/python3.11/lib-dynload/libssl.3.dylib: cannot load 'libssl.3.dylib' (load command 0x80000034 is unknown)
	/Applications/Get iPlayer Automator.app/Contents/Resources/python/lib/python3.11/lib-dynload/libssl.3.dylib: cannot load 'libssl.3.dylib' (load command 0x80000034 is unknown)
youtube-dl finished downloading
AppController: Downloads Finished

**Comment by skovatch** [comment]
@jeanpaulz85 Thanks, I'm trying another approach and will post another test soon.

**Reply by skovatch** [reply]
I have updated the test binary package. Thanks to @sleaux-meaux for the suggestion of yt-dlp. It's now a standalone app that doesn't rely on an external version of python, and is universal2 for M1 and Intel hardware.

**Comment by sleaux-meaux** [comment]
Worked on 10.15, but failed on 10.13 and 10.10 (Intel). GiA error dialog appears, but verbose  log showed nothing:
```
Downloading Trigger Point: Series 1
INFO: Preparing Request for Auth Info
INFO: Requesting Metadata.
DEBUG: Metadata response status code: 200
INFO: Metadata processed.
DEBUG: youtube-dl args:["https://www.itv.com/hub/trigger-point/10a0591a0001", "--user-agent", "Mozilla/5.0", "-f", "mp4/best", "-o", "/Users/xxx/Movies/TV Shows/Trigger Point - Series 1/Trigger Point - Series 1.s01e01.Episode 1.%(ext)s", "--verbose"]
DEBUG: youtube-dl environment: ["SSL_CERT_FILE": "/Applications/Get iPlayer Automator.app/Contents/Resources/cacert.pem", "PATH": "/Applications/Get iPlayer Automator.app/Contents/Resources/yt-dlp_macos:/Applications/Get iPlayer Automator.app/Contents/Resources/get_iplayer/utils/bin"]
youtube-dl finished downloading
AppController: Downloads Finished
```
Invoking from prompt showed:
```
$ export PATH="/Applications/Get iPlayer Automator.app/Contents/Resources/yt-dlp_macos:/Applications/Get iPlayer Automator.app/Contents/Resources/get_iplayer/utils/bin:$PATH"
$ export SSL_CERT_FILE="/Applications/Get iPlayer Automator.app/Contents/Resources/cacert.pem"
$ "/Applications/Get iPlayer Automator.app/Contents/Resources/yt-dlp_macos/yt-dlp_macos" "https://www.itv.com/hub/trigger-point/10a0591a0001" --verbose
Traceback (most recent call last):
  File "yt_dlp/__main__.py", line 16, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "yt_dlp/__init__.py", line 16, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "yt_dlp/options.py", line 15, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "yt_dlp/utils.py", line 33, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 495, in exec_module
  File "ssl.py", line 98, in <module>
ImportError: dlopen(/Applications/Get iPlayer Automator.app/Contents/Resources/yt-dlp_macos/lib-dynload/_ssl.cpython-39-darwin.so, 2): Library not loaded: @loader_path/../libssl.3.dylib
  Referenced from: /Applications/Get iPlayer Automator.app/Contents/Resources/yt-dlp_macos/lib-dynload/_ssl.cpython-39-darwin.so
  Reason: no suitable image found.  Did find:
	/Applications/Get iPlayer Automator.app/Contents/Resources/yt-dlp_macos/lib-dynload/../libssl.3.dylib: cannot load 'libssl.3.dylib' (load command 0x80000034 is unknown)
	/Applications/Get iPlayer Automator.app/Contents/Resources/yt-dlp_macos/lib-dynload/../libssl.3.dylib: cannot load 'libssl.3.dylib' (load command 0x80000034 is unknown)
[482] Failed to execute script '__main__' due to unhandled exception!
```
Perhaps try the prebuilt Python from python.org. That should guarantee the SSL module is compatible with its own OpenSSL libs. It worked for me, anyway.

**Reply by skovatch** [reply]
That error means that libssl.3.dylib wasn't built correctly -- it linked the shared library with a loader command that isn't recognized on macOS 10.13 and earlier. I used MacPorts to get openssl because it can build univesal2 libraries. But it looks like I forgot to set MACOS_DEPLOYMENT_TARGET=10.10 before building.

**Comment by skovatch** [comment]
I finally rebuilt libssl and libcrypto. The download link above now has a new version to try out. 
Unfortunately for me ITV seems to be blocking my smart DNS provider, so I'm hoping one of you can try it out on older macOS versions and let me know if it works.

**Comment by jeanpaulz85** [comment]
I got this error, I am using M1 MacOs 12.3

Jean-Paul 



Get iPlayer Automator 1.23b1 Initialized.
GetITVShows: ITV Cache Update Starting 
<AppController: 0x1506768c0>: Updating Programme Index Feeds...

/Applications/Get iPlayer Automator copy.app/Contents/Resources/get_iplayer/perl/bin/get_iplayer
-e14400
--type=tv
--refresh
--nopurge
--profile-dir=/Users/jeanpaulz85/Library/Application Support/Get iPlayer Automator
.*
GetITVShows (Info): Programmes: 517
INFO: Indexing tv programmes (concurrent)
INFO: Added 0 tv programmes to cache
INFO: 7855 matching programmes
GetITVShows (Info): Episodes: 6668
GetITVShows: Update Finished
<AppController: 0x1506768c0>: Index Updated.


INFO: Loading proxy settings...
INFO: No proxy to load
INFO: Proxy load complete.
INFO: No proxy will be used
AppController: Starting Downloads


Downloading Show 1/1:

Downloading Trigger Point: Series 1
INFO: Preparing Request for Auth Info
INFO: Requesting Metadata.
DEBUG: Metadata response status code: 200
INFO: Metadata processed.
DEBUG: youtube-dl args:["https://www.itv.com/hub/trigger-point/10a0591a0006", "--user-agent", "Mozilla/5.0", "-f", "mp4/best", "-o", "/Users/jeanpaulz85/Movies/TV Shows/Trigger Point/Trigger Point.s01e06.Episode 6.%(ext)s", "--write-sub", "-k", "--verbose"]
DEBUG: youtube-dl environment: ["PATH": "/Applications/Get iPlayer Automator copy.app/Contents/Resources/yt-dlp_macos:/Applications/Get iPlayer Automator copy.app/Contents/Resources/get_iplayer/utils/bin", "SSL_CERT_FILE": "/Applications/Get iPlayer Automator copy.app/Contents/Resources/cacert.pem"]
[debug] Command-line config: ['https://www.itv.com/hub/trigger-point/10a0591a0006', '--user-agent', 'Mozilla/5.0', '-f', 'mp4/best', '-o', '/Users/jeanpaulz85/Movies/TV Shows/Trigger Point/Trigger Point.s01e06.Episode 6.%(ext)s', '--write-sub', '-k', '--verbose']
[debug] Encodings: locale US-ASCII, fs utf-8, out ascii (No ANSI), err ascii (No ANSI), pref US-ASCII
[debug] yt-dlp version 2021.12.27 [6223f67a8] (mac_dir)
[debug] Python version 3.9.9 (CPython 64bit) - macOS-12.3-arm64-64bit
[debug] exe versions: ffmpeg 4.3.2-tessus
[debug] Optional libraries: Cryptodome, sqlite
[debug] Proxy map: {}
[ITV] 10a0591a0006: Downloading webpage
[debug] Using fake IP 25.106.10.155 (GB) as X-Forwarded-For
[debug] [ITV] Extracting URL: https://www.itv.com/hub/trigger-point/10a0591a0006
ERROR: [ITV] 10a0591a0006: Unable to download webpage: The read operation timed out (caused by timeout('The read operation timed out')); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output. (caused by timeout('The read operation timed out')); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output.
  File "yt_dlp/extractor/common.py", line 717, in _request_webpage
  File "yt_dlp/YoutubeDL.py", line 3458, in urlopen
  File "urllib/request.py", line 517, in open
  File "urllib/request.py", line 534, in _open
  File "urllib/request.py", line 494, in _call_chain
  File "yt_dlp/utils.py", line 1392, in https_open
  File "urllib/request.py", line 1350, in do_open
  File "http/client.py", line 1377, in getresponse
  File "http/client.py", line 320, in begin
  File "http/client.py", line 281, in _read_status
  File "socket.py", line 704, in readinto
  File "ssl.py", line 1241, in recv_into
  File "ssl.py", line 1099, in read
youtube-dl finished downloading
AppController: Downloads Finished

**Comment by skovatch** [comment]
I'm now getting 403 errors when yt-dlp starts the download, even though the same code worked a few days ago. This is a bit concerning... both youtube-dl and yt-dlp don't seem to be working for me from a unix shell either. I'm wondering if ITV is now blocking them altogether or it's just a short-term problem.

**Comment by sleaux-meaux** [comment]
Works on 10.10 and 10.13, though thumbnail embedding fails on all runs. This appears in logs:

```
No thumbnail URL, skipping
INFO: No thumbnail downloaded
WARN: AtomicParsley key: --artwork, value: nil
```

**Reply by skovatch** [reply]
That turns out to be a small bug I can take care of along with the embedded python. Unfortunately ITV is now broken for me, so I can't really test it.
